File: packages/plugin-sdk/assembly/badge.ts
import { peekoo_set_peek_badge } from "./host";
import { quote } from "./json";
import { writeString } from "./memory";
import { BadgeItem } from "./types";

/**
 * Replace all badge items for this plugin.
 * Pass an empty array to clear the badge.
 */
export function set(items: BadgeItem[]): void {
  // Manually build JSON array since AS JSON support is limited
  let json = "[";
  for (let i = 0; i < items.length; i++) {
    if (i > 0) json += ",";
    const item = items[i];
    json += "{\"label\":" + quote(item.label) + ",\"value\":" + quote(item.value);
    if (item.icon !== null) {
      json += ",\"icon\":" + quote(changetype<string>(item.icon));
    }
    if (item.countdown_secs > 0) {
      json += ",\"countdown_secs\":" + item.countdown_secs.toString();
    }
    json += "}";
  }
  json += "]";
  peekoo_set_peek_badge(writeString(json));
}

File: crates/peekoo-plugin-sdk/src/badge.rs
//! Peek badge overlay.
//!
//! Set badge items that are displayed on the desktop pet's peek overlay.
//! Requires the `notifications` permission.
//!
//! ```no_run
//! use peekoo_plugin_sdk::prelude::*;
//!
//! fn example() -> Result<(), Error> {
//!     peekoo::badge::set(&[
//!         BadgeItem {
//!             label: "Water".into(),
//!             value: "~5 min".into(),
//!             icon: Some("droplet".into()),
//!             countdown_secs: Some(300),
//!         },
//!     ])?;
//!     Ok(())
//! }
//! ```

use extism_pdk::Error;

use crate::host_fns::peekoo_set_peek_badge;
use crate::types::BadgeItem;

/// Replace all badge items for this plugin.
///
/// Pass an empty slice to clear the badge.
///
/// Internally serialises the items to a JSON string because the
/// `peekoo_set_peek_badge` host function accepts a raw `String`
/// (unlike the other host functions which use `Json<T>`).
pub fn set(items: &[BadgeItem]) -> Result<(), Error> {
    let json = serde_json::to_string(items)
        .map_err(|e| Error::msg(format!("badge::set serialize error: {e}")))?;

    unsafe {
        peekoo_set_peek_badge(json)?;
    }
    Ok(())
}
